Document new product page extendability (#1643)
PrestaShop/docs: modules/sample-modules/extend-product-page.md

---
title: Extending the new product page form
weight: 3
---

# Extending the new product page form {{< minver v="8.1.0" >}}

The new Back Office product page introduced in {{< minver v="8.1.0" >}} removed several hooks previously available: 

- `displayAdminProductsCombinationBottom`
- `displayAdminProductsSeoStepBottom`
- `displayAdminProductsShippingStepBottom`
- `displayAdminProductsQuantitiesStepBottom`
- `displayAdminProductsMainStepLeftColumnBottom`
- `displayAdminProductsMainStepLeftColumnMiddle`
- `displayAdminProductsMainStepRightColumnBottom`
- `displayAdminProductsOptionsStepTop`
- `displayAdminProductsOptionsStepBottom`
- `displayAdminProductsPriceStepBottom`

The only `displayAdminProduct*` hook that was not removed is: 

- `displayAdminProductsExtra`

{{% notice info %}}

> Altough conserved for Backwards compatibility, usage of this hook (`displayAdminProductsExtra`) is not recommended for new modules.

Reviewer: ```suggestion
Although kept for backward compatibility, this hook (`displayAdminProductsExtra`) is not recommended for new modules.
```

{{% /notice %}}

> In this guide, we will discover how to extend the product page by adding custom fields, in the old and new way of doing this.

Reviewer: ```suggestion
In this guide, we will discover how to extend the product page by adding custom fields, in the old and new ways of doing this.
```

> Finally, we will discover how to add a new Tab to the product page. 

Reviewer: ```suggestion
Finally, we will discover how to add a new tab to the product page, which is possible for a new product page from PrestaShop 8.1. 
```

`demooldhooks.php`: 

```php
declare(strict_types=1);

use Symfony\Component\Form\Extension\Core\Type\TextType;

class DemoOldHooks extends Module
{
  public function __construct()
  {
    // [...]
  }

  /**
   * @return bool
   */
  public function install()
  {
      return parent::install() && $this->registerHook(['displayAdminProductsSeoStepBottom']);
  }

  public function hookDisplayAdminProductsSeoStepBottom($params)
  {
    $productId = $params['id_product'];
    $formFactory = $this->get('form.factory');
    $twig = $this->get('twig');

    $product = new Product($productId);

    $form = $formFactory
      ->createNamedBuilder('seo_special_field', TextType::class, "")
      ->getForm();

    $template = '@Modules/demooldhooks/views/templates/seo_special_field.html.twig';

    return $twig->render($template, [
      'seo_special_field' => $form->createView()
    ]);
  }
}
```

`views/templates/seo_special_field.html.twig`: 

```php
<h3>SEO Special field</h3>
{{ form_widget(seo_special_field) }}
```

Before {{< minver v="8.1.0" >}}, that would produce: 

![custom field in SEO tab in older versions of PrestaShop](../img/old-product-form/seo-custom-field.png)

> From {{< minver v="8.1.0" >}}, that would produce nothing, since this hook (`displayAdminProductsSeoStepBottom`) is no more used. 

Reviewer: ```suggestion
From {{< minver v="8.1.0" >}}, this field won't be displayed since this hook (`displayAdminProductsSeoStepBottom`) is no longer available.
```

use PrestaShopBundle\Form\FormBuilderModifier;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\Form\FormBuilderInterface;

final class ProductFormModifier
{
    /**
     * @var FormBuilderModifier
     */
    private $formBuilderModifier;

    /**
     * @param FormBuilderModifier $formBuilderModifier
     */
    public function __construct(
        FormBuilderModifier $formBuilderModifier
    ) {
        $this->formBuilderModifier = $formBuilderModifier;
    }

    /**
     * @param int|null $productId
     * @param FormBuilderInterface $productFormBuilder
     */
    public function modify(
        int $productId,
        FormBuilderInterface $productFormBuilder
    ): void {
     
        $seoTabFormBuilder = $productFormBuilder->get('seo');
        $this->formBuilderModifier->addAfter(
            $seoTabFormBuilder, // the tab
            'tags', // the input/form from which to insert after/before
            'demo_module_custom_field', // your field name
            TextType::class, // your field type
            [
                'label' => 'SEO Special Field', // you can remove the label if you dont need it by passing 'label' => false

>                 'label_attr' => [ // customize label by any html attribute

Reviewer: ```suggestion
                'label_attr' => [ // customize label with any HTML attribute
```

                    'title' => 'h2',
                    'class' => 'text-info',
                ],
                'attr' => [
                    'placeholder' => 'SEO Special field',
                ],

>                 // this is just an example, but in real case scenario you could have some data provider class to wrap more complex cases

Reviewer: ```suggestion
                // this is just an example, but in real case scenario, you could have some data provider class to handle more complex cases
```

                'data' => "",
                'empty_data' => '',
                'form_theme' => '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig',
            ]
        );
    }
}
```

> This module is creating a Form Builder Modifier (`ProductFormModifier`), adding a `TextType` field to the `seo` tab form builder from the `ProductForm Builder`, added after the `tags` form element. 

Reviewer: ```suggestion
This module uses a Form Builder Modifier (`FormBuilderModifier `), and adds a `TextType` field to the `SEO` tab form builder from the `ProductForm Builder`,  after the existing `tags` form element. 
```

This Form Builder Modifier is hooked to the `actionProductFormBuilderModifier` hook. 

This produces this form: 

![custom field in SEO tab in newer versions of PrestaShop](../img/new-product-form/seo-custom-field.png)

{{% notice note %}}

> This new way of adding custom fields to the product page allows you for more precise positioning, you can now position your fields/forms exactly where you want. 

Reviewer: ```suggestion
This new way of adding custom fields to the product page allows you for more precise positioning. You can now position your fields/forms exactly where you want. 
```

| `displayAdminProductsShippingStepBottom` | Bottom of Shipping tab | `shipping` | after `carriers` |
| `displayAdminProductsQuantitiesStepBottom` | Bottom of Quantities tab | `stock` | after `low_stock_threshold` |
| `displayAdminProductsMainStepLeftColumnBottom` | Basic settings tab, under related product | `description` | after `related_products` |
| `displayAdminProductsMainStepLeftColumnMiddle` | Basic settings tab, under description | `description` | after `description` |
| `displayAdminProductsMainStepRightColumnBottom` | Basic settings tab, under create a category | `description` | after `categories` |
| `displayAdminProductsOptionsStepTop` | Top of Options tab | `options` | before `visibility` |
| `displayAdminProductsOptionsStepBottom` | Bottom of Options tab | `options` | after `product_suppliers` |
| `displayAdminProductsPriceStepBottom` | Bottom of Pricing tab | `pricing` | after `priority_management` |

Form Type details: [`EditProductFormType`](https://github.com/PrestaShop/PrestaShop/blob/8.1.0-beta.1/src/PrestaShopBundle/Form/Admin/Sell/Product/EditProductFormType.php)

> ## Extend a sub form

Reviewer: ```suggestion
## Extend a subform
```

> Sub forms can be extended aswell, for example, to add a new input on each combination of a product: 

Reviewer: ```suggestion
Subforms can be extended as well, for example, to add a new input on each combination of a product: 
```

    {
        $this->formBuilderModifier->addAfter(
            $combinationFormBuilder,
            'references',
            'demo_module_custom_field',
            TextType::class,
            [
                'label' => $this->translator->trans('Demo custom field', [], 'Modules.Demoproductform.Admin'),
                'label_attr' => [
                    'title' => 'h2',
                    'class' => 'text-info',
                ],
                'attr' => [
                    'placeholder' => $this->translator->trans('Your example text here', [], 'Modules.Demoproductform.Admin'),
                ],
                'data' => $customCombination->custom_field,
                'empty_data' => '',
                'form_theme' => '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig',
            ]
        );
    }
}
```

This example will add a `TextType` input on each `Combination` of a `Product`. 
A complete working example and implementation is available in our [example-module repository](https://github.com/PrestaShop/example-modules/tree/master/demoproductform).

## Add a custom Tab to Product Page

> {{< minver v="8.1.0" >}} introduced a new feature : Custom tabs in product page. 

Reviewer: ```suggestion
{{< minver v="8.1.0" >}} introduced a new feature: custom tabs on the product page. 
```

> A complete working example and implementation is available in our [example-module repository](https://github.com/PrestaShop/example-modules/tree/master/demoproductform).

Reviewer: ```suggestion
A complete working example of implementation is available in our [example-module repository](https://github.com/PrestaShop/example-modules/tree/master/demoproductform).
```

> - Modify the Product Form builder with `actionProductFormBuilderModifier` hook, and create a modifier: 

Reviewer: ```suggestion
- You can extend the product form builder with `actionProductFormBuilderModifier` hook with a created modifier: 
```

```yml
services:
    DemoNewHooks\Form\Type\CustomTabType:
        class: DemoNewHooks\Form\Type\CustomTabType
        parent: 'form.type.translatable.aware'
        public: true
        arguments:
            - '@=service("prestashop.adapter.data_provider.currency").getDefaultCurrency()'
        tags:
            - { name: form.type }
```

Add a `use` statement in your `ProductFormModifier`: 

```php
use DemoNewHooks\Form\Type\CustomTabType;
```

Finally, add this `CustomTabType` to the `$productFormBuilder` variable in your `modify` method in the `ProductFormModifier`: 

```php
/**
 * @param int|null $productId
 * @param FormBuilderInterface $productFormBuilder
 */
public function modify(
    int $productId,
    FormBuilderInterface $productFormBuilder
): void {
    
    $idValue = $productId ? $productId->getValue() : null;
    $customProduct = new CustomProduct($idValue);

    $this->formBuilderModifier->addAfter(
        $productFormBuilder,
        'pricing',
        'custom_tab',
        CustomTabType::class,
        [
            'data' => [
                'custom_price' => $customProduct->custom_price,
            ],
        ]
    );
}
```

{{% notice note %}}

> Any type added directly to the `$productFormBuilder` (and not to a sub tab) will be considered as a tab.

Reviewer: ```suggestion
Any type added directly to the `$productFormBuilder` (and not to a sub-tab) will be considered as a tab.
```

{{% /notice %}}

## Backwards compatibility for displayAdminProductsExtra hook

> A custom Form Type `ExtraModulesType` as been added to {{< minver v="8.1.0" >}}, to allow backwards compatibility for modules implementing `displayAdminProductsExtra` hook.

Reviewer: ```suggestion
A custom Form Type `ExtraModulesType` has been added to {{< minver v="8.1.0" >}}, to allow backward compatibility for modules implementing `displayAdminProductsExtra` hook.
```

> If a module has registered a hook on `displayAdminProductsExtra`, a custom tab will be added on the new product page, handled by `ExtraModulesType`. 

Reviewer: ```suggestion
If a module registers to `displayAdminProductsExtra` hook, a custom tab will be added on the new product page, handled by `ExtraModulesType`. 
```

Custom tab with all the modules that hook into this hook, right? Not a custom tab only for this one module?

{{% notice info %}}

> Altough conserved for Backwards compatibility, usage of this hook (`displayAdminProductsExtra`) is not recommended for new modules.

Reviewer: ```suggestion
Although kept for backward compatibility, this hook (`displayAdminProductsExtra`) is not recommended for new modules.
```
